libs/common/ring_buffer.hpp

```diff
@@ -0,0 +1,171 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_COMMON_RING_BUFFER_HPP
+#define IROHA_COMMON_RING_BUFFER_HPP
+
+namespace iroha {
+
+    template<typename T, size_t Count> class RingBuffer final {
+    public:
+        using type = T;
+        using handle = size_t;
+        using callback = std::function<void(handle, T&)>;
+
+    private:
+        static_assert(Count > 0, "Unexpected count value. It must be above 0.");
+        enum { kSizeType = sizeof(type) };
+
+        struct Node {
+            uint8_t data[kSizeType];
+            bool active;
```

Comment:
seems redundant to store this in every node. we can just have 2 handles - beginning and "past the end" of used area and compare given handle with them. there will be 2 cases, depending on whether the used area spans across the edges of the buffer: `***E....B***` and `...B******E...`. For the case when begin and end match, we need a bool `isEmpty`.

although this may seem too complex compared to storing bool for each node, this will not be used outside `assert`s, so the release code will be only better.

Reply:
I added this bool to keep state in the case we want to remove data from the middle of the buffer, if we want so. If we don't need this, than I can redesign it via 2 markers.
You don't need a bool isEmpty :)